---
title: "Early History of BioPerl"
layout: default
---

*Note: This article describes aspects of BioPerl up to 2010. Much work by a number of key individuals since then is not accounted for.*

Our group owes its early organizational support to its association with the award-winning VSNS-BCD BioComputing Courses; some students of the 1996 course (Chris Dagdigian, Richard Resnick, Lew Gramer, Alessandro Guffanti, and others) have contributed code and commentary. Georg Fuellen, the VSNS-BCD chief organizer was one of the driving forces behind bioperl. Steven Brenner, who was an early adopter of Perl for bioinformatics provided some of the early work on BioPerl. Lincoln Stein has long provided guidance and code.

Bioperl was then taken up by people developing code at the large genome centres. In particular at Stanford, Steve Chervitz at Stanford, Ian Korf at the Washington University Genome Sequencing Centre (St Louis) and at the Sanger Centre (Cambridge UK) Ewan Birney. All of the originalC code XS extensions were provided by Ewan Birney, although additional work has been provided by Aaron Mackey (Bio::SeqIO plugins for chromatogram formats). BioPerl is used in production at these sites, indicating that is both useful and that it works.

Jason Stajich and Hilmar Lapp joined BioPerl for the drive towards a 0.7 release over 2000 and the first part of 2001, which includes a revised feature location model, richer feature objects (in particular genes) and more and better tools. Peter Schatner and Lorenz Pollak contributed serious chunks of code, being the AlignIO and bptutorial scripts and the BPLite port to bioperl respectively. At this time BioPerl was being used in absolute earnest by the Ensembl group which shook out a number of problems in the code base. Additional compatibility with the GenQuire/Sequence Workbench (GUI package) (Mark Wilkinson and David Block) and BioCORBA (Jason Stajich, Brad Chapman and Alan Robinson) and finally Game-XML (Brad Marshall) provided more interoperability.

In May 2001 Heikki Lehvxc3xa4slaiho who has been a very significant coder, bugfixer and contributor was made part of the BioPerl "core" group. Members of the BioPerl core are responsible for getting new releases out the door and generally handling all the behind the scenes work that running an open source project involves.

The first ever Open Bio Hackathon co-sponsored by O'Reilly & Associates and Electric Genetics held in two parts in Tucson, Arizona, USA and Cape Town, South Africa in February and March of 2002. Large strides to combine the work of all the Open Bioinformatics Foundation projects. At the retreat Lincoln Stein was invited to become part of the BioPerl core to assist in setting vision and standards for the project. Lincoln's projects relating to BioPerl include Gbrowse, a component in the GMOD project, and major components and in the BioPerl toolkit.

In May of 2002 , Bioperl 1.0, a significant release, was made available. This release represented a large stride in our attempt to produce a stable and easy-to-use toolkit. Bug fix releases since then have improved the toolkit and its reliability in a production environment. August 2002 saw the Open Bioinformatics Foundation's BOSC 2002 annual conference for open source software development.

In October 2002 we published a paper about the project in addition to a paper by Lincoln Stein and collegues about the Gbrowse genome browser built with BioPerl. A full BioPerl publications page is available illustrating how the toolkit is used.

In March 2002 we held a Hackathon in Tucson, AZ (part of O'Reilly Bioinformatics Conference) and in Capetown, South Africa. Some pictures and some press 1 2 from the events. The hackathon was sponsored by O'Reilly & Associates and Electric Genetics. AstraZeneca and Dalke Scientific Software also sponsored part of the event. In Feb 2003 we held a Hackathon in Singapore sponsored by Apple Asia.

Jon Orwant at The Perl Journal gave us permission to reprint Lincoln Stein's article on How Perl Saved the Human Genome Project. He also worked with the Perl Institute (now the Perl Mongers) to arrange our `bio.perl.org` DNS entry.

One of our first generation server hardware for http://bioperl.org (and other open-bio.org hosted projects) was provided by Compaq Computer Corporation. The donation was facilitated by both the Pharmaceutical Sales and High Performance Technical Computing (HPTC) groups.

The OBF servers reside in Cambridge, Massachusetts USA with colocation facilities and internet bandwidth donated by Wyeth Research. In particular Dr. Steven Howes, Kenny Grant & Rich DiNunno have made significant efforts on our behalf.

Today, BioPerl is a thriving open source project, with nearly all of the founding developers still involved. As of mid-2010, the BioPerl mailing lists process an average volume of 9 messages per day.
